Tool for Dataset Normalization

kamehta2 opened the pull request:
Implementing Issue #5 

- Examine dataset to see if it will benefit from normalization
- Add a user option to select normalization if found useful

Work is still in progress

kamehta2 committed:
Tool for dataset normalization

Implementing enhancement issue #5

- Examine dataset to see if it will benefit from normalization
- Add a user option to select normalization if found useful

Work is still in progress

kamehta2 committed:
Refactored code to include Class NormCheck for re-usability and ease of usage

Added Class NormCheck
Added two new functions

Tested code with sample inputs

kamehta2 commented:
I will add some docstrings to my code.

kamehta2 committed:
Adding Docstring to every function

kamehta2 committed:
read the data from "main", not in function

rohitchadaram commented on hs628_tools/normalization_tool.py:
```diff
@@ -0,0 +1,67 @@
+        The user input gives choice to user if they want to proceed with normalization or not.
+        If data does not need normalization, the function returns with an informative message reflecting the same.
+        """
+        if percent_diff < self.cutoff:
```
Can you explain more how you arrived at this cut-off value ? and also what does the percent difference really signify ? to decide if you want to do normalization

rohitchadaram commented on hs628_tools/normalization_tool.py:
```diff
@@ -0,0 +1,67 @@
+
+if __name__ == '__main__':
+    import doctest
+    doctest.testmod()
```
You do not need this doctest as you do not really have testcases in your docstrings.

EHWUSF replied:
I myself would emphasize the flip side of this-- you have started to add some support for unit testing here, so it would be great to go ahead and add at least a couple simple doctests above someplace.

kamehta2 replied:
The reason I added this section was because I wanted to add some unit tests. However, it is tricky to add tests because of return value as ndarray.
I will still attempt to add a couple of unit tests. If I am unable to, i will definitely remove this section.

RoxanneXin commented on hs628_tools/normalization_tool.py:
```diff
@@ -37,7 +35,6 @@ def get_input(self, percent_diff):
         else:
```
Really like your idea! And your code is well structured and you explain it clearly enough to understand. 
For the cutoff = 0.3, can I ask where it's come from and why?

kamehta2 replied:
Cutoff is based on analyzing the mean ratios right now. Mean ratio is actually equal to "minimum mean divided by maximum mean, multiplied by 100". Based on the analysis of values found, closer to 0 will mean that the difference in mean values is really large, and we want to avoid such a scenario. To narrow down to the value of 0.3 as the cutoff, I checked different datasets, and used different cutoffs, and after all the analysis, came up with this value. But I definitely know that depending on dataset, the optimum value of cut-off may change, and hence I have also left this as a user choice, where user has the ability to modify the default cut-off.

EHWUSF commented on hs628_tools/normalization_tool.py:
```diff
@@ -0,0 +1,67 @@
+
+
+class NormCheck:
+    def __init__(self, alldata, cutoff=0.3, alldata1=None):
```
can you think of a more descriptive variable name for ```alldata1```  (and it would  be best to describe the attributes in a docstring also).

EHWUSF commented on hs628_tools/normalization_tool.py:
```diff
@@ -0,0 +1,67 @@
+    def find_percent_diff(self):
+        """ 
+        This function takes the data as a numpy.ndarray. It calculates mean for each column.
+        It further processes this data to return the percentage difference.
```
I am having a hard time seeing the calculation being done here as a 'percentage difference', please clarify that.

Also, I think that comparing the feature means may not be so helpful-- two features could both have means close to zero even if their respective total ranges differ by a factor of a thousand.  That may not seem likely to happen by random, but someone could have decided to previously recenter their data on zero without having rescaled it. I think that feature range (range ratio) or variation/standard-deviation should probably be considered (along with the mean).

kamehta2 replied:
Yes, what I have done can be better named as "mean ratio", which is actually equal to "minimum mean divided by maximum mean, multiplied by 100". The mean values used in the formula are found after inspecting all features.

I definitely agree that range ratio will be a better way to go with, compared to mean ratio. I will see how i can incorporate range ratio along with the mean ratio.

EHWUSF commented on hs628_tools/normalization_tool.py:
```diff
@@ -0,0 +1,67 @@
+        allmean = self.alldata1.mean(axis=0)
+        allmeanmin = allmean.min()
+        allmeanmax = allmean.max()
+        percent_diff = allmeanmin * 100 / float(allmeanmax)
```
beware of dividing by a value that could be zero.

kamehta2 replied:
thanks for pointing that out. I will definitely add a check for the "allmeanmax" variable.

EHWUSF commented on hs628_tools/normalization_tool.py:
```diff
@@ -0,0 +1,67 @@
+        percent_diff = allmeanmin * 100 / float(allmeanmax)
+        return percent_diff
+
+    def get_input(self, percent_diff):
```
This function is broadly OK at least from an intermediate programming standpoint, but here are a number of points that would make it better.
 - it is named simply 'get_input', but it doesn't always get input, and it seems like the most important thing it does is decide whether or not to recommend normalization.   I'd recommend separating those actions, but if kept together a name like 'decide_normalization' or somesuch seems more aptly descriptive.
- somewhat related to the above, the return value is ambiguous.  The "Yes", "No" are fairly clear, but the "This dataset may not need normalization." return value seems a bit odd.  It is not ever presented for the user, so doesn't need to be a long string, which invites programming/maintenance errors if some other part of the code were to try to check for that value.  Also, suppose that the program did recommend normalization, but then instead of "Yes" or "No", the user entered "This dataset may not need normalization."   the rest of your program would have no way of detecting that.  If you did maintain the dual nature of this function, I'd very strongly recommend a return value of None to indicate that the normalization question was not even presented to the user.
- you don't really present any information from your analysis to help the user make their decision, so why ask them at all?  I think it would be best to either present some information to assist with the decision or just require that they make a decision rule in advance (by either accepting or overriding your default value for the threshold). (And I would strongly lean towards the latter approach-- you could also return or output some information about the analysis you did, in case the user does want to make an evaluation.

kamehta2 replied:
I will review the flow again, including the name of the function. I also will update the usage instructions accordingly.

EHWUSF commented on hs628_tools/normalization_tool.py:
```diff
@@ -0,0 +1,67 @@
+        This function takes the data as a numpy.ndarray. It calculates mean for each column.
+        It further processes this data to return the percentage difference.
+        """
+        rowwise_missing_entries = len(alldata) - np.sum(alldata == alldata, axis=0)
```
This checks for np.nan, but doesn't check for other ways that data might be indicated as missing.  At the very least, you'd need to document that to the user.  It might be better to just require that the user provide data without any missing entries, because otherwise you also have to have a way to let the user know exactly which columns you've removed.

kamehta2 replied:
I will be adding user instructions that will ask for data that does not have any missing values.

EHWUSF commented on hs628_tools/normalization_tool.py:
```diff
@@ -0,0 +1,67 @@
+        if rescale.upper() == "YES":
+            alldata_normalized = preprocessing.normalize(self.alldata1)
+            alldata_normalized = alldata_normalized
+            return alldata_normalized
```
As noted above, it could easily cause trouble for the user if this is missing columns from what they provided.  If you are going to stick with that aspect of the approach, you would need to have a programmatic way to provide that information to the calling code (not just reporting out to the user).

kamehta2 replied:
I will be adding user instructions that will ask for data that does not have any missing values. Hence, this should not be an issue anymore.

EHWUSF commented on hs628_tools/normalization_tool.py:
```diff
@@ -0,0 +1,67 @@
+        else:
+            return "This dataset may not need normalization."
+
+    def compare(self, rescale):
```
Although 'compare' is one part of what this function currently does, for a user writing code that uses your module, does that name really help clarify what this method is doing?

kamehta2 replied:
Since I changed the flow a couple of times, the names definitely don't do proper justice to what the function is completely doing. This function returns normalized data to the user. I will rename it so that it is clear.

EHWUSF commented on hs628_tools/normalization_tool.py:
```diff
@@ -0,0 +1,67 @@
+            alldata_normalized = alldata_normalized
+            return alldata_normalized
+        else:
+            return "You have selected not to normalize your dataset."
```
Returning either an ndarray or a string is an odd sort of possible return values.  A more typical treatment (and one more readily handled by the calling code) would be to return either an ndarray or None.  Plus, this return string is not really always accurate-- it could very well be the case that the user made no selection at all because your code decided there is no need to ask them.

kamehta2 replied:
I will prefer to change the return value to "None" for the else case. I was trying to make it more informative for the user, however, I will now cover some of the usage guidelines in a better way in the user instructions document, and make the code not have return values that are strings.

EHWUSF commented on hs628_tools/normalization_tool.py:
```diff
@@ -0,0 +1,67 @@
+    import doctest
+    doctest.testmod()
+
+    alldata = np.genfromtxt("College.csv", delimiter=',', skip_header=1)
```
you have a hardcoded file reference here, but that file is not currently part of the repository.  At a minimum you should do one or more of the following: (i) just add the test file into the repository as part of this branch, (ii) add comments to let someone know where to download the file, and/or (iii) add some checking or a try:except clause in your code so that it uses the local file if available or otherwise directly accesses a remote copy of the file using a URL.

kamehta2 replied:
I can add the file into repository. Not sure yet about "iii", although it seems like a good idea.

EHWUSF commented on hs628_tools/normalization_tool.py:
```diff
@@ -0,0 +1,67 @@
+    
+    percentage = normalization.find_percent_diff()
+    rescale_input = normalization.get_input(percentage)
+    normalized_data = normalization.compare(rescale_input)
```
If not programmatically checking the test results, you should at least add comments for the user about the expected  test results.

kamehta2 replied:
Will do. Am still planning to update doctests - or else I will update more comments as well as update usage instructions for expected test results.

kamehta2 committed:
Submitting changes suggested

Change functions and attributes name
Add function to calculate range ratio
Check if the denominator is zero
Combining two functions into one as decide_normalization
No return value as a string

kamehta2 committed:
Including cutoff value for range_ratio

kamehta2 committed:
Adding documentation for the final commit
